Manage dependencies using Poetry

jleaniz opened the pull request:
### Description of the change
This pull request moves dependency management to a new ```pyproject.toml``` file to be used with Poetry.

Dependencies were cleaned up where possible. Dockerfiles have been edited to use poetry to install dependencies for unit testing. In addition, the production Dockerfiles will now install Turbinia and its dependencies via PyPi instead of ```setup.py install```. 

A new GitHub action was added to automatically push new release packages to PyPi containing a wheel and source. This will allow the Cloud Build triggers to use ```pip install turbinia==$RELEASE_TAG``` to install the packages. The ```publish-pypi``` action will trigger once a new release and tag are published in GitHub.

### Applicable issues

<!-- Enter any applicable Issues here (You can reference an issue using #) -->
- fixes #1311  
- fixes #733 

### Additional information

<!-- If there's anything else that's important and relevant to your pull request, mention that information here.-->

### Checklist

<!-- [Place an '[X]' (no spaces) in all applicable fields. Please remove unrelated fields.] -->

- [ ] All tests were successful.
- [ ] Unit tests added.
- [ ] Documentation updated.

jleaniz commented:
E2E tests will currently fail because the ```RELEASE_TAG``` does not exist yet.

jleaniz commented:
@rbdebeer-zz Please take a look at the approach and let me know your thoughts.

hacktobeer commented on docker/tests/Dockerfile:
```diff
@@ -0,0 +1,153 @@
+RUN set -x \
+  && echo "Compiling Yara from source..."
+
+# Fetch and compile libyara
```
this and the actual fraken build can be replaced by:
COPY --from=us-docker.pkg.dev/osdfir-registry/turbinia/release/fraken:latest` /bin/fraken /opt/fraken/fraken

when https://github.com/google/turbinia/pull/1353 is in

jleaniz replied:
done

hacktobeer commented on docker/server/Dockerfile:
```diff
@@ -13,15 +13,12 @@ RUN apt-get update && apt-get -y upgrade && apt-get -y install \
-    && cd /tmp/ && python3 setup.py install
+# Use the REELASE_TAG argument passed via --build-arg cli argument
+ARG RELEASE_TAG
+RUN pip3 install turbinia==$RELEASE_TAG
```
this will be replaced with the corresponding poetry command (in all Dockerfiles) I assume.

jleaniz commented on AUTHORS:
```diff
@@ -5,8 +5,9 @@ Turbinia is developed by (in alphabetical order):
-* Wajih Yassine (wajihyassine@gmail.com)
 * Johan Berggren (jberggren@gmail.com)
+* Juan Leaniz (juan.leaniz@gmail.com)
+* Wajih Yassine (wajihyassine@gmail.com)
```
of course!

jleaniz commented:
@hacktobeer PTAL at latest changes. I adjusted a few things. Included dockerignore files, using poetry to install dependencies in the containers, updated start.sh scripts, etc.

hacktobeer commented:
@jleaniz would love to see this PR reviewed and merged, a lot of the (docker/dev) work moving forward will depend on the new poetry changes and I do not want to merge bigger changes with this big PR.

jleaniz commented:
> @jleaniz would love to see this PR reviewed and merged, a lot of the (docker/dev) work moving forward will depend on the new poetry changes and I do not want to merge bigger changes with this big PR.

It's ready for review, please take a look

hacktobeer commented on .github/workflows/publish-pypi.yml:
```diff
@@ -0,0 +1,107 @@
+
+  publish-to-pypi:
+    name: Publish Python 🐍 distribution 📦 to PyPI
+    if: startsWith(github.ref, 'refs/tags/')  # only publish to PyPI on tag pushes
```
I don't think we push any other tags then the release tags but should someone push *any* tag it will push a release to pypi. Would it be safe to assume our tags always start with '20' and filter on 'refs/tags/20' to prevent a accidental push when someone tags eg dev-expiremental or whatever at some point....if too paranoid please ignore!

hacktobeer reviewed (approved):
LGTM, only small nits as I already reviewed before. 
Request: 
- Update the SRE release documentation on how we first change the version in pyproject.toml before github tagging a release
- Update the contribution page on how you need to lock the poetry dependency versions and include those in a PR if you need a new requirement. https://turbinia.readthedocs.io/en/latest/developer/contributing.html

aarontp commented on docker/worker/Dockerfile:
```diff
@@ -76,61 +26,33 @@ RUN apt-get update && apt-get -y upgrade && apt-get -y install \
-RUN useradd -r -s /bin/nologin -G disk,sudo -u 999 turbinia
+RUN useradd -r -s /sbin/nologin -G disk,sudo -u 999 turbinia
 RUN echo "turbinia ALL = (root) NOPASSWD: ALL" > /etc/sudoers.d/turbinia
+RUN echo "Defaults secure_path=\"/venv/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\"" >> /etc/sudoers.d/turbinia
```
Is this the right path?  Does it need to be /home/turbinia/.venv/bin or similar?

aarontp replied:
Do we need /opt/fraken or other dependency paths here?

jleaniz replied:
Good catch. Should be ```/home/turbinia/.venv```
I added ```/opt/fraken``` just in case

aarontp reviewed (approved):
LGTM, this will be great to have!

Can you add some documentation somewhere on how to update the poetry.lock file if that doesn't already exist?  Thanks!
